File: typescript/src/service/exchangeRateService.ts
import {CurrencyAmount} from '../model/currencyAmount'

export const BASE_CURRENCY: string = 'CHF'

export interface ExchangeRateService {
     convert(currencyAmount: CurrencyAmount, toCurrency: string): CurrencyAmount;
}

export class ExchangeRateServiceImpl implements ExchangeRateService {

    private readonly baseCurrency: string = 'CHF';
    private readonly exchangeRates: Map<string, number> = new Map<string, number>([
        ['EUR', 1.0744],
        ['USD', 0.9103],
        ['GBP', 1.1794]
    ]);

    convert(currencyAmount: CurrencyAmount, toCurrency: string): CurrencyAmount {
        return new CurrencyAmount(toCurrency,
            currencyAmount.amount * this.exchangeRate(currencyAmount.currency, toCurrency));
    }

   private exchangeRate(fromCurrency: string, toCurrency: string): number {
        if (fromCurrency == toCurrency) {
            return 1;
        }
        if (fromCurrency == this.baseCurrency) {
            return 1 / this.requireExchangeRate(toCurrency);
        }
        if (toCurrency == this.baseCurrency) {
            return this.requireExchangeRate(fromCurrency);
        }
        return this.exchangeRate(fromCurrency, this.baseCurrency) * this.exchangeRate(this.baseCurrency, toCurrency);
    }

   private requireExchangeRate(currency: string) {
        const exchangeRate = this.exchangeRates.get(currency);
        if (!exchangeRate) {
            throw new Error(`unsupported currency: ${currency}`);
        }
        return exchangeRate;
    }
}

File: python/service/exchangerateservice.py
from models.currencyamount import CurrencyAmount
from util.check import required, is_currency


class ExchangeRateService:

    def __init__(self):
        self.base_currency = 'CHF'
        self.precision = 2
        self.exchange_rates = {
            'EUR': 1.0744,
            'USD': 0.9103,
            'GBP': 1.1794
        }

    def convert(self, currency_amount: CurrencyAmount, to_currency: str):

        required('currency_amount', currency_amount)
        required('to_currency', to_currency, is_currency())

        return CurrencyAmount(to_currency,
                              round(currency_amount.amount * self.exchange_rate(currency_amount.currency, to_currency),
                                    self.precision))

    def exchange_rate(self, from_currency: str, to_currency: str):
        required('from_currency', from_currency, is_currency())
        required('to_currency', to_currency, is_currency())

        if from_currency == to_currency:
            return 1
        if from_currency == self.base_currency:
            return 1 / self.require_exchange_rate(to_currency)
        if to_currency == self.base_currency:
            return self.require_exchange_rate(from_currency)
        return self.exchange_rate(from_currency, self.base_currency) * \
               self.exchange_rate(self.base_currency, to_currency)

    def require_exchange_rate(self, currency: str):
        required('currency', currency, is_currency())
        if currency not in self.exchange_rates:
            raise Exception(f'unsupported currency: {currency}')
        return self.exchange_rates[currency]
